src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cassert>
+#include <set>
+#include <sstream>
+
+#include "tracer.hpp"
+
+using namespace std;
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    thread_local std::shared_ptr<CTracer> tracer = nullptr;
+    std::shared_ptr<CTracer> CreateTracer(int preferredLayerDuration)
+    {
+        tracer = std::make_shared<CTracer>(preferredLayerDuration);
+        return tracer;
+    }
+    std::shared_ptr<CTracer> CreateTracer(int preferredLayerDuration, const std::unordered_map<OpId, std::string>& opNames)
+    {
+        tracer = std::make_shared<CTracer>(preferredLayerDuration, opNames);
+        return tracer;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer's ISimulator implementation
+    //------------------------------------------------------------------------------------------------------------------
+    Qubit CTracer::AllocateQubit()
+    {
+        size_t qubit = qubits.size();
+        qubits.emplace_back(QubitState{});
+        return reinterpret_cast<Qubit>(qubit);
+    }
+    void CTracer::ReleaseQubit(Qubit /*qubit*/)
+    {
+        // nothing for now
+    }
+    std::string CTracer::QubitToString(Qubit q)
+    {
+        size_t qubitIndex = reinterpret_cast<size_t>(q);
+        const QubitState& qstate = this->UseQubit(q);
+
+        stringstream str(std::to_string(qubitIndex));
+        str << " last used in layer " << qstate.layer << "(pending zero ops: " << qstate.pendingZeroOps.size() << ")";

```

Comment:
This feels like a functionality alien for `QubitToString()`. I would expect from `QubitToString()` just to convert the qubit descriptor to string, and nothing else.
I would expect the other info to be added by some decorator outside of `QubitToString()`.

Reply:
What do you mean by "qubit descriptor"? The qubit's id? `QubitToString` is a public API on ISimulator -- the _only_ way for the client to get information about the qubit. I suppose, we could do effectively nothing here and provide a more expressive printout functionality as public API on CTracer (e.g. different verbosity levels, output to json, etc).

Reply:
Sorry, with "qubit descriptor" I meant the instance of the `Qubit` type - the parameter `q`.

Reply:
The Qubit type is meaningless, it's just a handle to QubitState. I actually don't know what a user might expect to be printed here... I think we'd need to see end-to-end usage scenarios before deciding on this.